Bluetooth: hci: Prevent naming conflict in hci_ecc.c

jori-nordic opened the pull request:
It seems JWT is using the same `default_CSPRNG` name as Bluetooth
in hci_ecc. Rename it, make it static and register it explicitly instead
of relying on global symbols.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
Bluetooth: hci: Prevent naming conflict in hci_ecc.c

It seems JWT is using the same `default_CSPRNG` name as the bluetooth
function. Rename it, make it static and register it explicitly instead
of relying on global symbols.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

Thalley reviewed (commented):
Searching for `default_CSPRNG` in the Zephyr codebase, I see that we have a 2 places in `tests` that also use this default name. Do they "need" to be fixed too? 

And there are even more in `modules` and `bootloader`.

jori-nordic committed:
jwt: set uECC csprng function only once

This (along with the parent commit) is to prevent a scenario where e.g.,
Bluetooth hci_ecc is in the middle of an uECC call, gets pre-empted by
`jwt_sign`, which then changes the uECC csprng function.

The same could happen the other way around, if the caller of `jwt_sign`
is a pre-emptable thread with a lower priority.

This way, the last setter of the uECC CSPRNG will be the one that is
always used.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic commented on subsys/bluetooth/host/hci_ecc.c:
```diff
@@ -347,13 +347,15 @@ void bt_hci_ecc_supported_commands(uint8_t *supported_commands)
 }
 
-int default_CSPRNG(uint8_t *dst, unsigned int len)
+static int bt_CSPRNG(uint8_t *dst, unsigned int len)
```
`modules` and `bootloader` results don't seem relevant. The only real user seems to be mcuboot and it shouldn't need a json library (jwt) or bluetooth support imo.
Actually, I think I can just change the `jwt` name to something else.

Ah I think I found a problem: the `jwt_sign` function calls `uECC_set_rng()` on every call. With the hci_ecc thread being pre-emptible, that could cause problems down the line. I have made it init only once in the next commit.

jori-nordic commented:
We discussed with @alwa-nordic that it'd be nice to have a system-wide CSPRNG that uECC would use, so multiple users can just use that instead of re-defining their own. Tagging @dleach02 for his input.

sjanc commented on subsys/bluetooth/host/hci_ecc.c:
```diff
@@ -347,13 +347,15 @@ void bt_hci_ecc_supported_commands(uint8_t *supported_commands)
 
 void bt_hci_ecc_init(void)
 {
+	uECC_set_rgn(&ecc_csprng);
```
typo...  uECC_set_rng()

also, as this configures global, could affect system in unexpected way

alwa-nordic reviewed (changes requested):
I tagged [area: Crypto / RNG](https://github.com/zephyrproject-rtos/zephyr/labels/area%3A%20Crypto%20%2F%20RNG) to get input on safety and framework.

I believe it may be unsafe to use JWT and Bluetooth together as they are.

Observations:
- The name `default_CSPRNG` implies uECC needs a cryptographically secure RNG.
- JWT uses TinyCrypt CTR-PRNG `tc_ctr_prng`, seeded with `sys_rand32_get`. This is a CSPRNG seeded with "not a cryptographically secure random number value". **RED FLAG** (Consider using `sys_csrand_get`.)
- JWT calls `uECC_set_rng` in its code in `jwt_sign`. It will override the RNG source every time `jwt_sign` is called.

Since the uECC library is stateful, it requires some cooperation between its users. I feel it should be the responsibility of Zephyr to define a policy here.

To ensure proper crypto in Bluetooth, I think we need either:
 - That Bluetooth calls `uECC_set_rgn` (after disabling interrupts?) before using uECC functions. Though I have one worry: If, first, uECC used for JWT with insecure RNG. Then, `uECC_set_rgn` sets a secure RNG. Is there some state left in uECC from the insecure RNG that can be an attack vector on Bluetooth?
 - That `uECC_set_rgn` is patched out of uECC, and `default_CSPRNG` is controlled by subsys random, which provides it with proper CSRNG.
 - A front-end API for uECC that will multiplex the library in some safe way.

jori-nordic commented on tests/crypto/tinycrypt/src/ecc_dsa.c:
```diff
@@ -611,7 +611,7 @@ int montecarlo_signverify(int num_tests, bool verbose)
 }
 
-int default_CSPRNG(uint8_t *dest, unsigned int size)
+int early_prng(uint8_t *dest, unsigned int size)
```
deduced from the first comment in this function. I assume `at this point in time` means something better might come later.

Thalley replied:
Sounds like an assumption to me :D "at this point in time" can also be a simple restriction, rather than something that is improved later :)

jori-nordic commented:
Closing as per @alwa-nordic 's comment, this PR was just a surface fix, but we need a real solution. @joerchan will look into that.

joerchan commented:
@jori-nordic Can you close if that is the conlusion?

jori-nordic commented:
> @jori-nordic Can you close if that is the conlusion?

Yep sorry didn't realize I had hit 'comment' instead of 'close with comment'
